### PR TITLE
struct bulk insert

### PR DESCRIPTION
**Pull request by petersunbag**
`InsertInto` and `InsertIntoForTag` support bulk insert

### CODE REVIEW

**Comment by huandu** [comment]
It sounds reasonable to let `Struct#InsertInto` support bulk insert. Thanks for your contribution.

I read your commit. I think it's better to change signature of related methods to `InsertInto(table string, value ...interface{})` and `InsertIntoForTag(table string, tag string, value ...interface{})` instead of using `value` as a struct or a slice of structs at the same time. I hope you can change the signature and implement this feature again.

The philosophy of the package design is to keep everything straight-forward without any magic. It's easy to understand that one `interface{}` is one piece of data, not many, and `value ...interface{}` or `[]interface{}` is a slice of data. If we treat one `interface{}` as one or many at the same time in a method like `InsertInto`, one will get confused and have to read doc carefully to understand the usage of `interface{}` in all methods. Obviously, if one has to do so to use a package correctly, the package is not straight-forward to use.

BTW, the CI failure is a bug in my travis ci config. I'll fix it soon. You can rebase your PR to my fix later.

**Comment by huandu** [comment]
A side note for how to use `Struct#InsertInto` if the signature is changed.

```go
// A global variable.
myStruct := sqlbuilder.NewStruct(new(MyStruct))

// Prepare a slice of MyStruct.
data := []*MyStruct{...}
ib := myStruct.InsertInto("foo", sqlbuilder.Flatten(data)...) // Use Flatten to expand data explicitly.
sql, args := ib.Build()
```

**Comment by huandu** [comment]
I've updated `.travis.yml` to fix build break on pull requests. Please rebase your PR to the latest commit and submit again.

**Comment by petersunbag** [comment]
Thanks for the _"straight-forward"_ sharing.

File: struct.go [hunk 1]
```diff
@@ -198,24 +200,34 @@ func (s *Struct) InsertIntoForTag(table string, tag string, value interface{}) *
...
+			vs = append(vs, v)
+		}
+	}
+	if len(vs) == 0 {
 		return ib
 	}
 
 	cols := make([]string, 0, len(fields))
-	values := make([]interface{}, 0, len(fields))
+	values := make([][]interface{}, len(vs), len(vs))
```
Comment (huandu) [inline struct.go]: Should simplify this statement to `values := make([][]interface{}, len(vs))`

File: struct.go [hunk 2]
```diff
@@ -198,24 +200,34 @@ func (s *Struct) InsertIntoForTag(table string, tag string, value interface{}) *
...
 
 	for _, f := range fields {
-		name := s.fieldAlias[f]
-		data := v.FieldByName(name).Interface()
 		cols = append(cols, f)
-		values = append(values, data)
+		name := s.fieldAlias[f]
+		for i, v := range vs {
+			data := v.FieldByName(name).Interface()
```
Comment (huandu) [inline struct.go]: The `data` should be set outside the for loop as it doesn't change in the loop at all.

File: struct.go [hunk 3]
```diff
@@ -198,24 +200,34 @@ func (s *Struct) InsertIntoForTag(table string, tag string, value interface{}) *
 		return ib
 	}
 
-	v := dereferencedValue(value)
+	var vs []reflect.Value
```
Comment (huandu) [inline struct.go]: Should use `make` for better performance.

`vs := make([]reflect.Value, 0, len(value))`

File: struct.go (hunk 2 above)
Comment (huandu) [reply struct.go]: I find my previous comment is not correct. You can ignore it without taking any action.

**Comment by iambudi** [comment]
**Question:**

Is there any option to ignore struct without db tag as part of insert? I need to separate header and items as two insert queries.

```go
type SalesOrder struct {
   ID int32 `db:"id"`
   CustomerID int32 `db:"customer_id"`
   OrderDate time.Time `db:"order_date"`
   Items: []*SalesOrderItem // <-- No tags here, should be ignored from query builder
}

type SalesOrderItem struct {
   ID int32 `db:"id"`
   SalesOrderID int32 `db:"sales_order_id"`
   ProductID int32 `db:"product_id"`
   Qty int32 `db:"qty"`
   Price float32 `db:"price"`
}
```
ib will generate the header query as `INSERT INTO sales_order (id, customer_id, order_date, Items)`. Here `Items` can be excluded.

**Comment by iambudi** [comment]
Solved. i should use InsertIntoForTag with `fieldtag`.

**Pull request merged**